pkg/pattern/metric/evaluator.go

```diff
@@ -0,0 +1,306 @@
+package metric
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+
+	loki_iter "github.com/grafana/loki/v3/pkg/iter"
+	"github.com/grafana/loki/v3/pkg/logproto"
+	"github.com/grafana/loki/v3/pkg/logql"
+	"github.com/grafana/loki/v3/pkg/logql/syntax"
+	"github.com/grafana/loki/v3/pkg/storage/chunk/cache/resultscache"
+)
+
+// TODO(twhitney): duplication with code in NewStepEvaluator
+func extractMetricType(expr syntax.SampleExpr) (Type, error) {
+	var typ Type
+	switch e := expr.(type) {
+	case *syntax.VectorAggregationExpr:
```

Comment:
do you have some example queries in mind? I do think this area could use more testing, but it's definitely an area my knowledge is weaker.

Reply:
You could with a binary operation. But let's not worry about this for now.

Reply:
Ha, TIL that one cannot embed range vectors:
```
fist_over_time(count_over_time({cluster="prod-eu-west-0", namespace="loki-prod"}[$__auto]))
```

Reply:
> You could with a binary operation. But let's not worry about this for now.

Yes, and these are explicitly not supported on this endpoint yet. It will error if you try.